lib/checks/mysql.js

```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+const mysql = require('mysql');
+const Check = require('../check');
+
+module.exports = class Mysql extends Check {
+    constructor(config) {
+        super(config);
+        this.error = new Error(`MySQL connection to ${config.url} failed.`);
+    }
+
+    async start () {
+        return await new Promise((resolve => {
+            const connection = mysql.createConnection({
+                host : this.config.url,
+                user : this.config.user,
+                password : this.config.password,
+                database : this.config.database,
+            });
+            connection.ping(err => {
+                if (err) {
+                    this.error = err;
+                    this.crit();
+                } else {
+                    this.ok();
+                }
+                connection.end();
+                setTimeout(this.start.bind(this), this.interval);
```

Comment:
Curious why the check sets a timeout and then tries again. Shouldn't the check just execute when the status check is executed by the framework which would happen when MGE calls down to the service?

Reply:
It's a bit different approach here - the status check of a service (`self-check`) and its first-level dependencies is performed by every service itself either in intervals (eg. `mysql`, `http` or `self`checks) or in reaction to  events (specific for implementation  - eg. `redis` or `mongo` check). Then resulting status is cached by the service until next loop or event occurs.
That way the overall API request for status can be significantly faster cause actual health check requests are distributed through the whole hierarchy of services. This can optimize the API reaction time especially in case of large ecosystem of services.
Does it make sense for you?

Reply:
True. On the other hand all setup is eventually consistent as MGE also has an interval parameter for checking the health of chosen service. Resuming there can be some situations when some failures can be hidden in any case with the respect to MGE interval value. 
I guess normally the interval values are going to be single seconds and so more significant failures which would need some reaction will pop up in MGE anyway. The shorter outages, when hidden because of system self cure of any kind won't matter that much, would thay?

Reply:
Even when we leave MGE for a while and think of invoking the API directly, we've got the eventual consistency anyway, as after a single API call returning whatever result this result can be invalid afyer a short while.
Resuming - the typical and probably the only one pragmatic use case is to perform the API calls in some interval to get the system health check view as a function of time, right?

Reply:
Yeah I think the usage we have is slightly different. We use MGE as a debugging tool and use the Health Checks framework on each service node with Sensu for alerting. We set this ping interval outside of MGE and the health checks framework with a tool called Sensu. We check/alert for outages based on the interval in Sensu and the result from the health checks `/status/aggregate?type=[internal|external]` endpoint. We tend not to cache any results in the service as we see that as inaccurate. As you point out, if you declare the interval value in the framework to be smaller than the interval in MGE, then it should mostly be OK. For simplicity reason we don't do this. I tend to think that a status check should be super fast and simple. If it is slow, then that in general is overall reflection of the dependency you are checking and the normal operation of that dependency in your service. Caching the result will improve the status check performance yes, but has the potential to mask other performance related issues of the service itself.

Does this make sense?